crates/sui-network/src/state_sync/mod.rs

```diff
@@ -0,0 +1,854 @@
+        self.tasks.spawn(task);
+    }
+}
+
+async fn notify_peers_of_checkpoint(
+    network: anemo::Network,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+    checkpoint: VerifiedCheckpoint,
+) {
+    let futs = peer_heights
+        .read()
+        .unwrap()
+        .heights
+        .iter()
+        // Filter out any peers who we know already have a checkpoint higher than this one
+        .filter(|(_peer_id, &height)| Some(checkpoint.sequence_number()) > height)
+        .map(|(peer_id, _height)| peer_id)
+        // Filter out any peers who we aren't connected with
+        .flat_map(|peer_id| network.peer(*peer_id))
+        .map(StateSyncClient::new)
+        .map(|mut client| {
+            let request = Request::new(checkpoint.inner().clone()).with_timeout(DEFAULT_TIMEOUT);
+            async move { client.push_checkpoint_summary(request).await }
+        })
+        .collect::<Vec<_>>();
+    futures::future::join_all(futs).await;
+}
+
+async fn get_latest_from_peer(peer: anemo::Peer, peer_heights: Arc<RwLock<PeerHeights>>) {
+    let peer_id = peer.peer_id();
+    let request = Request::new(GetCheckpointSummaryRequest::Latest).with_timeout(DEFAULT_TIMEOUT);
+    let response = StateSyncClient::new(peer)
+        .get_checkpoint_summary(request)
+        .await
+        .map(Response::into_inner);
+    update_peer_height(&peer_heights, peer_id, &response);
+}
+
+fn update_peer_height(
+    peer_heights: &RwLock<PeerHeights>,
+    peer_id: PeerId,
+    response: &Result<Option<Checkpoint>, Status>,
+) {
+    match response {
+        Ok(latest) => {
+            peer_heights
+                .write()
+                .unwrap()
+                .update_peer_height(peer_id, latest.clone());
+        }
+        Err(status) => {
+            trace!("get_latest_checkpoint_summary request failed: {status:?}");
+            peer_heights.write().unwrap().heights.remove(&peer_id);
+        }
+    }
+}
+
+async fn query_peers_for_their_latest_checkpoint(
+    network: anemo::Network,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+    tx: mpsc::WeakSender<StateSyncMessage>,
+) {
+    let peer_heights = &peer_heights;
+    let futs = peer_heights
+        .read()
+        .unwrap()
+        .heights
+        .keys()
+        // Filter out any peers who we aren't connected with
+        .flat_map(|peer_id| network.peer(*peer_id))
+        .map(|peer| {
+            let peer_id = peer.peer_id();
+            let mut client = StateSyncClient::new(peer);
+
+            let request =
+                Request::new(GetCheckpointSummaryRequest::Latest).with_timeout(DEFAULT_TIMEOUT);
+            async move {
+                let response = client
+                    .get_checkpoint_summary(request)
+                    .await
+                    .map(Response::into_inner);
+                update_peer_height(peer_heights, peer_id, &response);
+                response
+            }
+        })
+        .collect::<Vec<_>>();
```

Comment:
same question re: can the `collect` be skipped

Reply:
same answer unfortunately.